Add PyTorch TensorIterator update post

kurtamohler opened the pull request:
This PR adds a new blog post that contains much of the information from @v0dro's `TensorIterator` post, but with updated code examples according to the recent changes to `TensorIterator`. I also added a bit more detail in the cumulative sum example. I did not duplicate the TH iterator history section from the original post, since that has not changed.

I also created a repo (https://github.com/kurtamohler/pytorch-TensorIterator-examples) that contains a standalone C++ pytorch app that runs the example code and checks for correctness. I haven't set up any sort of nightly tests for it, but I'm planning on doing that so we'll automatically know when any of these APIs change again.

cc @rgommers

kurtamohler commented on posts/2020/04/pytorch-tensoriterator-internals.md:
```diff
@@ -10,6 +10,12 @@
+The history section of this post is still relevant, but `TensorIterator`'s
+interface has changed significantly. For an update on the new API, please check
+out [this new blog
+post](../../2021/04/pytorch-tensoriterator-internals-update.html).
```
This link doesn't seem to work. I need to find out the correct syntax.

mattip replied:
Maybe `/blog/2021/04/pytorch-tensoriterator-internals-update.html`

mattip replied:
Also, [nikola](https://www.getnikola.com/handbook.html#doc) suggests you can do 
```
:doc:`pytorch-tensoriterator-internals-update.`
```

rgommers commented:
Thanks Kurt! @nikitaved would you like to review the described changes for the new `TensorIterator` API?

> I also created a repo (https://github.com/kurtamohler/pytorch-TensorIterator-examples) that contains a standalone C++ pytorch app that runs the example code and checks for correctness. I haven't set up any sort of nightly tests for it, but I'm planning on doing that so we'll automatically know when any of these APIs change again.

That seems like a good idea. Weekly should work too:)

nikitaved commented on posts/2021/01/pytorch-tensoriterator-internals-update.md:
```diff
@@ -0,0 +1,264 @@
+
+# Basics of TensorIterator and TensorIteratorConfig
+
+In order to create a `TensorIterator`, a `TensorIteratorConfig` must be created
+first. `TensorIteratorConfig` specifies the input and output tensors that will
```
A fun or a not so fun fact, `TensorIteratorConfig` is not copyable nor copy-assignable :(

nikitaved commented on posts/2021/01/pytorch-tensoriterator-internals-update.md:
```diff
@@ -0,0 +1,264 @@
+Iterations using `TensorIterator` can be classified as point-wise iterations or
+reduction iterations. This plays a fundamental role in how iterations using
+`TensorIterator` are parallelized - point-wise iterations can be freely
+parallelized along any dimension and grain size while reduction operations have
+to be either parallelized along dimensions that you're not iterating over or by
+performing bisect and reduce operations along the dimension being iterated.
```
You can still perform reductions while parallelizing over a full Tensor. Take `sum`, for example. It is all about whether you can create an index map from the input to the output. In the case of sum you can `unsqueeze` the reduction dimension with stride 0 for the output, and set `output.size(dim) = input.size(dim)`. The kernel stays the same: `*output_ptr += *input_ptr`.

And, in general, not including the reduction dimension into the loop is not efficient for CUDA.

nikitaved replied:
Sorry, only if this operation is atomic. Please, ignore.

nikitaved commented on posts/2021/01/pytorch-tensoriterator-internals-update.md:
```diff
@@ -0,0 +1,264 @@
+
+auto iter = iter_config.build();
+
+// Copies data from input into output
+auto copy_loop = [](char** data, const int64_t* strides, int64_t n) {
+  auto* out_data = data[0];
+  auto* in_data = data[1];
+
+  for (int i = 0; i < n; i++) {
+    // assume float data type for this example
+    *reinterpret_cast<float*>(out_data) += *reinterpret_cast<float*>(in_data);
```
The comment indicates the copy operation, while the loop is doing the sum.

nikitaved commented on posts/2021/01/pytorch-tensoriterator-internals-update.md:
````diff
@@ -0,0 +1,264 @@
+});
+```
+
+Writing the kernel in this way ensures that the value returned by the lambda
+passed to `cpu_kernel` will populate the corresponding position in the target
+output tensor.
````
Only if `input` strictly broadcasts over `output`. That is `input` broadcasts over `output` and `output.shape >= input.shape`.

nikitaved commented on posts/2021/01/pytorch-tensoriterator-internals-update.md:
```diff
@@ -0,0 +1,264 @@
+at::Tensor result = at::empty_like(self);
+
+at::TensorIteratorConfig iter_config;
+auto iter = iter_config
+  .check_all_same_dtype(false)
+  .resize_outputs(false)
```
Not needed here, probably. But if you restride your tensors with zero stride, you will need to use `.set_check_mem_overlap(false)`, because Tensors with zero strides are considered having overlaps in TensorIterator.

kurtamohler replied:
`resize_outputs` is required when using `declare_static_shape`. I get this error if I don't include it:

```
terminate called after throwing an instance of 'c10::Error'
  what():  resize_outputs() must be called before declare_static_shape(...)
Exception raised from declare_static_shape at /pytorch/aten/src/ATen/TensorIterator.cpp:100
```

nikitaved replied:
The comment was about the memory overlap check, sorry for the confusion.

nikitaved reviewed (commented):
Thank you very much, @kurtamohler ! I really enjoyed this update.
I have just minor comments.

nikitaved commented on posts/2021/01/pytorch-tensoriterator-internals-update.md:
```diff
@@ -0,0 +1,264 @@
+Iterations using `TensorIterator` can be classified as point-wise iterations or
+reduction iterations. This plays a fundamental role in how iterations using
+`TensorIterator` are parallelized - point-wise iterations can be freely
+parallelized along any dimension and grain size while reduction operations have
+to be either parallelized along dimensions that you're not iterating over or by
+performing bisect and reduce operations along the dimension being iterated.
+Parallelization can also happen using vectorized operations. Parallelization
```
Relevant for CUDA: normally you still parallelize over the reduction dimension, but then you will need synchronizations (atomics for example) to avoid race conditions.

kurtamohler commented:
Thanks for all the helpful comments @nikitaved! I will push an update within the next couple days

Lezcano commented on posts/2021/01/pytorch-tensoriterator-internals-update.md:
```diff
@@ -0,0 +1,264 @@
+be properly understood in order to use `TensorIterator` effectively.
+
+When performing a reduction operation (see the `sum_out` code in
+[ReduceOps.cpp](https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/ReduceOps.cpp#L433)),
```
This line has changed (and probably those in the other links as well). In this case, the new correct line is 517, but perhaps you want to want to point to `make_reduction` in `ReduceOpsUtil.h`, as it looks like they refactored `sum_out`.

rgommers replied:
`blob/master` should always be `blob/<commit-hash>`

Lezcano commented on posts/2021/01/pytorch-tensoriterator-internals-update.md:
```diff
@@ -0,0 +1,264 @@
+  for (int64_t i = 0; i < n; ++i) {
+    // Calculate cumulative sum for each element of the vector
+    auto cumulative_sum = (at::acc_type<float, false>) 0;
+    for (int64_t i = 0; i < self_dim_size; ++i) {
```
This `i` shadows the one declared 3 lines ago. Should it be a `j`?

kurtamohler commented:
Thanks for all the help! I've incorporated suggestions, fixed links, and I also added a new section "Helper functions". I thought it would be a good idea to include a code example using "make_reduction" since it's very commonly found throughout PyTorch

kurtamohler commented:
I also set up weekly actions for my example code repo, so we should know if any changes ever break them: https://github.com/kurtamohler/pytorch-TensorIterator-examples/actions

kurtamohler commented on posts/2021/02/pytorch-tensoriterator-internals-update.md:
````diff
@@ -0,0 +1,321 @@
+iter.for_each(loop);
+```
+
+### Helper functions
````
Hey @nikitaved, could you take a look at this new section when you get a chance? Thanks!

trallard commented:
Task linked: [CU-cgxq18 Blog post on PyTorch TensorIterator - Kurt (in progress)](https://app.clickup.com/t/cgxq18)

nikitaved commented on posts/2021/02/pytorch-tensoriterator-internals-update.md:
```diff
@@ -0,0 +1,321 @@
+
+  for (int64_t i = 0; i < n; i++) {
+    // assume float data type for this example
+    *reinterpret_cast<float*>(out_data) += *reinterpret_cast<float*>(in_data);
+    in_data += strides[1];
+  }
```
Shouldn't `out_data` advance as well?

kurtamohler replied:
I've added an explanation of why `strides[0] == 0`

nikitaved commented on posts/2021/02/pytorch-tensoriterator-internals-update.md:
````diff
@@ -0,0 +1,321 @@
+iter.for_each(sum_reduce_loop);
+```
+
+# Conclusion
+
+This post was a very short introduction to what `TensorIterator` is actually
+capable of. If you want to learn more about how it works and what goes into
+things like collapsing the tensor size for optimizing memory access, a good
````
I think it will be of much help to later on cover the CUDA-related TensorIterator parts.

kurtamohler replied:
I agree that it would be good to eventually cover CUDA. But maybe it would be best to do that in another blog post to make things more easily digestible?

nikitaved replied:
Sure, I totally agree!

nikitaved commented on posts/2021/02/pytorch-tensoriterator-internals-update.md:
```diff
@@ -0,0 +1,321 @@
+int64_t dim = 1;
+bool keepdim = false;
+
+// `make_reduction` will allocate result Tensor for us, so we
+// can leave it undefined
+at::Tensor result;
```
What will happen if we feed a defined Tensor? Is it going to be resized or we can expect a segfault if it happens to be too small?

Lezcano replied:
I believe that the correct treatment of `out=` arguments could make for a fantastic addition to the post, as it is surprisingly tricky at time.

kurtamohler replied:
`make_reduction` resizes the result tensor if it's defined and isn't the correct size. I can add a note about this--perhaps it would be instructive if I init `result` with an incorrect size in the example.

You're right that overlaps aren't being checked, so I'll also add a note about that as well.

Lezcano replied:
Ivan Yashchuk wrote a number of PRs making sure that the behaviour of the `out=` version was consistent across `torch.linalg` functions. It believe that it's done by using something like `at::native::resize_output(result, sizes);`. This is done by `TensorIterator` automatically unless one specifies `  .resize_outputs(false)`, as it is done in the other example. I do not know whether `make_reduction` handles this correctly, but by Nik's comment I assume it does not (?).

Lezcano replied:
While doing a review today I got into the `make_reduction` code and I found that it does not handle it well (it does not raise a warning if it resizes an `out` tensor that was fed with the wrong size. Hopefully that will be solved soon (perhaps in that same PR), so perhaps it's fine not to touch on that here, or simply mention that it is handled internally in `make_reduction`, and otherwise one would need to use `.resize_outputs` or `at::native::resize_output`.

kurtamohler replied:
Ok, since `make_reduction`'s handling of out tensors may change soon, I agree that it's probably best not to attempt to cover it in detail here

kurtamohler commented:
I think this is ready, let me know what you think @Lezcano and @nikitaved

Lezcano commented:
LGTM!

kurtamohler commented:
Does anyone know about the `netlify/quansight-labs/deploy-preview` CI job? It doesn't seem like the failure is my fault, but just want to make sure

mattip commented:
circleci does not merge master into a PR, so you could clear that via rebasing/merging master into your branch.

mattip commented:
The post is not visible in the netlify preview because of this line in the `conf.py`. Over at hpyproject.org they set this to True. https://github.com/Quansight-Labs/quansight-labs-site/blob/11b18d18f85a8955ddd7c110ba27790790a7eb7e/conf.py#L939

kurtamohler commented:
@mattip, I updated the time to be in the past, but the netlify preview still doesn't show it. It also doesn't show the warning I added to the old TensorIterator post

mattip commented:
Does `make build` work locally?

kurtamohler commented:
Yeah, it builds and I can see the post on the site with `nikola server --browser`. However, on the `/blog/` page which lists all the blog posts (this page on the public site: https://labs.quansight.org/blog/), I see the entire TensorIterator blog post--it's not cut off with "Read more..." for some reason.

mattip commented:
I see the [this post](https://854-179573943-gh.circle-artifacts.com/0/output/blog/2021/04/pytorch-tensoriterator-internals-update/index.html) and the updated [older post](https://854-179573943-gh.circle-artifacts.com/0/output/blog/2020/04/pytorch-tensoriterator-internals/index.html) on the circleci build. I also see [the post](https://deploy-preview-156--quansight-labs.netlify.app/blog/2021/04/pytorch-tensoriterator-internals-update/) on the deploy preview and the [old post modified](https://deploy-preview-156--quansight-labs.netlify.app/blog/2020/04/pytorch-tensoriterator-internals/index.html).

I don't know why the post does not show up on the front page.

kurtamohler commented:
@mattip, it looks like the netlify preview works now with my last commit. Maybe it requires the TEASER_END for it to show up on `/blog/`?

kurtamohler commented:
Hey @nikitaved, I think this post is ready to merge unless you have any more thoughts

nikitaved reviewed (approved):
Great work, @kurtamohler ! Looking forward to seeing it go live!

rgommers commented:
Awesome. I'll merge it tomorrow.

rgommers commented on posts/2021/04/pytorch-tensoriterator-internals-update.md:
```diff
@@ -0,0 +1,337 @@
+<!--
+.. title: PyTorch TensorIterator Internals (TensorIteratorConfig Update)
```
rendering of this title comes out awkwardly: 

<img width="569" alt="image" src="https://user-images.githubusercontent.com/98330/114194420-c1ee0600-994f-11eb-87c8-583e9aa43fbd.png">

`TensorIteratorConfig` is also too specific for being in the title I think. How about _"PyTorch TensorIterator internals - 2021 update"_?

kurtamohler replied:
I like that title, I've changed it

rgommers commented on posts/2021/04/pytorch-tensoriterator-internals-update.md:
```diff
@@ -0,0 +1,337 @@
+In April 2020, Sameer Deshmukh wrote a blog article discussing
+[PyTorch TensorIterator Internals](/blog/2020/04/pytorch-tensoriterator-internals/index.html). Recently,
+however, the interface has changed significantly. This post describes how to
+use the current interface as of January 2021. Much of the information from the
```
just checking - still valid as of April 2021?

kurtamohler replied:
Good catch, I changed it to April

rgommers reviewed (approved):
LGTM now, thanks @kurtamohler and all reviewers!
